### ISSUE TITLE
Findings while using `createCollections`

### ISSUE THREAD

**Issue by noxify** [souporserious/mdxts#163]
Hi,

currently testing the `createCollection`  functionality (with version 1.6.2 ) and found the following:

> Note: I know the docs aren't finished and `createCollection` and their sub methods are currently under development. Just want to share what I have seen :) 

If you need a repro, please let me know :) 

> Fun fact: after starting to build it, I have seen that you have created something similar with the `examples/design-system` :)

## Error: listen EADDRINUSE: address already in use :::5996

While using

``` 
"dev": "concurrently \"mdxts watch\" \"next dev\"",
```

I always get the following error message after starting the dev server:

```
❯ pnpm dev

> mdxts-docs-template@0.1.1 dev /Users/marcusreinhardt/Development/mdxts-docs-template
> concurrently "mdxts watch" "next dev"

[1]   ▲ Next.js 14.2.5
[1]   - Local:        http://localhost:3000
[1] 
[1]  ✓ Starting...
[1]    automatically enabled Fast Refresh for 1 custom loader
[1] Error: listen EADDRINUSE: address already in use :::5996
[1]     at Server.setupListenHandle [as _listen2] (node:net:1902:16)
[1]     at listenInCluster (node:net:1959:12)
[1]     at Server.listen (node:net:2061:7)
[1]     at new WebSocketServer (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/ws@8.18.0/node_modules/ws/lib/websocket-server.js:102:20)
[1]     at new WebSocketServer (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/mdxts@1.6.2_next@14.2.5_@babel+core@7.25.2_react-dom@18.3.1_react@18.3.1__react@18.3.1__prett_656yul4n2kdfwmj64r4niyucfq/node_modules/mdxts/dist/cjs/src/project/rpc/server.js:22:24)
[1]     at createServer (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/mdxts@1.6.2_next@14.2.5_@babel+core@7.25.2_react-dom@18.3.1_react@18.3.1__react@18.3.1__prett_656yul4n2kdfwmj64r4niyucfq/node_modules/mdxts/dist/cjs/src/project/server.js:17:20)
[1]     at nextConfig.webpack (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/mdxts@1.6.2_next@14.2.5_@babel+core@7.25.2_react-dom@18.3.1_react@18.3.1__react@18.3.1__prett_656yul4n2kdfwmj64r4niyucfq/node_modules/mdxts/dist/cjs/src/next/index.js:55:46)
[1]     at Object.webpack (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/@next+mdx@14.2.3_@mdx-js+loader@2.3.0_webpack@5.93.0_/node_modules/@next/mdx/index.js:35:29)
[1]     at getBaseWebpackConfig (/Users/marcusreinhardt/Development/mdxts-docs-template/node_modules/.pnpm/next@14.2.5_@babel+core@7.25.2_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/build/webpack-config.js:1825:32)
[1]     at async Promise.all (index 1) {
[1]   code: 'EADDRINUSE',
[1]   errno: -48,
[1]   syscall: 'listen',
[1]   address: '::',
[1]   port: 5996
[1] }
[1]  ✓ Ready in 3.1s
[1]  ○ Compiling / ...
```

## Missing frontmatter export

I have the following collection configuration:

```ts
import type { FileSystemSource, MDXContent } from "mdxts/collections"
import { createCollection } from "mdxts/collections"

export interface DocSchema {
  default: MDXContent
  frontmatter: {
    title: string
    summary: string
    tags?: string[]
  }
}

export type DocsSource = FileSystemSource<DocSchema>

export const DocsCollection = createCollection<DocSchema>(
  "@/content/docs/**/*.{ts,tsx,mdx}",
  {
    title: "Docs",
    baseDirectory: "docs",
    basePath: "docs",
  },
)
```

In the `page.tsx` I use the following code:

```ts
  const source = DocsCollection.getSource(params.slug)

  // if we can't find the source then return a 404
  if (!source) {
    return notFound()
  }
  const sections = await source.getSources({ depth: 1 })

  const exports = await source.getNamedExport("frontmatter").getValue()
  const Content = await source.getDefaultExport().getValue()
```

The completion shows the correct types if I hover over the exports const:
![image](https://github.com/user-attachments/assets/7207a901-57ac-4eed-97ba-789b3cde7685)

But if I try to log the result via `console.log({exports})`, I just get an `undefined`.

The content `document.mdx` has the following structure:

```md
---
title: Build a Button Component in React
date: 2024-03-01
summary: Learn how to build a reusable Button component in React that can be used across your application.
tags:
  - react
  - design systems
---

In modern web development, creating reusable UI components is a must for efficiency and scalability. React, with its component-based architecture, allows developers to build encapsulated components that manage their own state and can be reused throughout applications. In this tutorial, we'll create a simple, reusable `Button` component using React.
```

## Missing headings

Previously we had access to the headings of a document - currently I can't see and property/method which provides this kind of information

## `tree` helper is missing

This is/was a nice helper, because we could use it to build the navigation based on the information.
Current workaround for me is to use something like this - Do you plan to provide an helper like this or is it now in our responsibility?

> Note: For me, it's fine, since every structure is different and needs other information, so providing a generic helper couldn't fit some use cases )
> 
> Maybe providing a code snippet to show how we can create our own tree is enough ( feel free to use the following snippet if it's helpful :) 

```
import type { DocSchema } from "@/config/collections"
import type {
  CollectionSource,
  FileExports,
  FileSystemSource,
} from "mdxts/collections"
import { DocsCollection } from "@/config/collections"

interface TreeItem {
  title: string
  path: string
  slug: string[]
  order: string
  depth: number
  children?: TreeItem[]
}

async function getTree<T extends FileExports>({
  input,
  maxDepth = 2,
  fromSource = false,
}: {
  input: CollectionSource<T> | FileSystemSource<T>[]
  maxDepth?: number
  fromSource?: boolean
}): Promise<TreeItem[]> {
  let sources: FileSystemSource<T>[]
  if (fromSource) {
    sources = input as FileSystemSource<T>[]
  } else {
    sources = await (input as CollectionSource<T>).getSources({ depth: 1 })
  }

  const tree: TreeItem[] = []
  for (const source of sources) {
    const treeItem = {
      title: source.getTitle(),
      path: source.getPath(),
      slug: source.getPathSegments(),
      order: source.getOrder(),
      depth: source.getDepth(),
      children:
        source.getDepth() <= maxDepth
          ? await getTree<T>({
              input: await source.getSources({ depth: 1 }),
              maxDepth,
              fromSource: true,
            })
          : [],
    }
    tree.push(treeItem)
  }
  return tree
}

const tree = await getTree<DocSchema>({ input: DocsCollection })
console.log({ tree })
```

**Comment by souporserious** [comment]
Thank you for trying out the new collections utility! I really appreciate all of the feedback ❤️ 

> Error: listen EADDRINUSE: address already in use :::5996

Hm did this happen immediately? Or if you kill the port will it work the first time? Wondering if I picked a bad static port number or it never closes properly 😅 

> Missing frontmatter export

This is intentional, at least right now. I'm debating whether or not to offer a codemod and have folks move their front matter to an export that can be statically analyzed more easily and not require extra processing. Is this something you would be willing to do?

So instead of this:

```mdx
---
title: Build a Button Component in React
date: 2024-03-01
summary: Learn how to build a reusable Button component in React that can be used across your application.
tags:
  - react
  - design systems
---
```

You'd write:

```mdx
export const frontmatter = {
  title: "Build a Button Component in React",
  date: "2024-03-01",
  summary: "Learn how to build a reusable Button component in React that can be used across your application.",
  tags: ["react", "design systems"]
};
```

The codemod would help convert existing MDX files to this format. Alternatively, a remark plugin can be used to do this for you, see the design system example [here](https://github.com/souporserious/mdxts/blob/main/examples/design-system/next.config.mjs#L9). This is what was done previously, but there is a caveat having to load every module to do things like sorting since I don't know that's an exported variable statically.

Front matter is statically analyzable as well, but breaks the new collections architecture of working with exports. There could be a special case for front matter specifically since it's so popular.

> Missing headings

The solve here will be exporting a headings variable using a remark plugin. Right now they are all tied together in the `mdx-plugins` package export, I want to split these up so you can use them granularly.

> tree helper is missing

The new architecture was built around making it much easier to build tree navigations yourself with more flexibility. Now you can do something like this which I'll have guides for:

```tsx
import Link from 'next/link'

import { PostsCollection, type PostSource } from '@/collections'

async function TreeNavigation({ Source }: { Source: PostSource }) {
  const Sources = await Source.getSources(1)
  const depth = Source.getDepth()
  const path = Source.getPath()
  const frontMatter = await Source.getNamedExport('frontmatter')
    .getValue()
    .catch(() => null)

  if (Sources.length === 0) {
    return (
      <li style={{ paddingLeft: `${depth}rem` }}>
        <Link
          href={path}
          style={{
            display: 'grid',
            color: 'white',
          }}
        >
          {frontMatter?.title || Source.getTitle()}
        </Link>
      </li>
    )
  }

  const childrenSources = Sources.map((ChildSource) => (
    <TreeNavigation key={ChildSource.getPath()} Source={ChildSource} />
  ))

  if (depth > 0) {
    return (
      <li style={{ paddingLeft: `${depth}rem` }}>
        <Link href={path} style={{ color: 'white' }}>
          {frontMatter?.title || Source.getTitle()}
        </Link>
        <ul
          style={{
            listStyle: 'none',
            padding: 0,
            margin: 0,
          }}
        >
          {childrenSources}
        </ul>
      </li>
    )
  }

  return (
    <ul
      style={{
        listStyle: 'none',
        padding: 0,
        margin: 0,
      }}
    >
      {childrenSources}
    </ul>
  )
}

export default async function BlogLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <div style={{ display: 'grid', gridTemplateColumns: '1fr 3fr' }}>
      <aside>
        <h2>Posts</h2>
        <ul
          style={{
            display: 'grid',
            gap: '1rem',
            listStyle: 'none',
            padding: 0,
            margin: 0,
          }}
        >
          {(await PostsCollection.getSources(1)).map((Source) => (
            <TreeNavigation key={Source.getPath()} Source={Source} />
          ))}
        </ul>
      </aside>
      <main>{children}</main>
    </div>
  )
}
```

**Comment by noxify** [comment]
Hi,

thanks for the detailed reply.

**frontmatter**:

* Now the "getNamedExport" functions makes sense for me :D 
* Using `export const frontmatter = {}` is fine for me and shouldn't be a problem in a new project
* Providing a codemod to convert the old schema to the new one would help a lot ( in our internal documentation, we have currently 328 documents and it would be a pain if we have to do this manually :D )

> There could be a special case for front matter specifically since it's so popular.

The question is: Is this what you want as creator? ( I assume you have a vision for the mdxts package ;) )
Personally, I think it's ok to say "we do not support the old frontmatter structure, use the codemod to convert to the new structure". 
As user, it would hurt the first time, but then it shouldn't be a problem at all.

> Note: I assume that other users have a different perspective here - As developer, I don't care if I have to write
> `---` or `export const frontmatter`.


**tree helper**:

> The new architecture was built around making it much easier to build tree navigations yourself with more flexibility. Now you can do something like this which I'll have guides for

That's perfect and should be enough to have an example to integrate it.

**mdxts cli problems**:

> Hm did this happen immediately? Or if you kill the port will it work the first time? Wondering if I picked a bad static port number or it never closes properly

Based on the terminal logs, it occurs immediately:

```
> concurrently "mdxts watch" "next dev"

[1]   ▲ Next.js 14.2.5
[1]   - Local:        http://localhost:3000
[1] 
[1]  ✓ Starting...
[1]    automatically enabled Fast Refresh for 1 custom loader
[1] Error: listen EADDRINUSE: address already in use :::5996
[1]     at Server.setupListenHandle [as _listen2] (node:net:1902:16)
[1]     at listenInCluster (node:net:1959:12)
```

I ensured that the ports isn't used before I ran the command via `lsof -i :5996`.
After I ran `pnpm dev`, the respone was:

```
❯ lsof -i :5996
COMMAND  PID            USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    4347 marcusreinhardt   15u  IPv6 0x5ed09b19923760ea      0t0  TCP *:5996 (LISTEN)
```

tested also `mdxts next` --> same result.

If I cancel the process, the port is also closed.

Based on the error, it seems that something is trying to start the mdxts server again ( maybe just a missing `isRunning()`check? )

...a few moments later...

i checked the `design-systems` example again and after installing `@next/mdx` ( & some other plugins ) and updating the import for `createMDXPlugin` the error is gone, but got now the `createContext can't be used in a server component` ( added also the webpack configuration to my next config ).

To make it easier, I published the repo: https://github.com/noxify/mdxts-docs-template

**Comment by noxify** [comment]
If I use 

```
import { createMdxtsPlugin } from "mdxts/next"
```

then the `createContext` error is gone. After updating the dev script to `"dev": "next dev"`, the `port already in use` error is also gone and the port is in use, but the `.mdxts` directory isn't created and I get the following error:

```
You can fix this error by taking the following actions:
- Make sure the ".mdxts" directory was successfully created and your tsconfig.json is aliases "mdxts" to ".mdxts/index.js" correctly.
- Make sure the file pattern is formatted correctly and targeting files that exist.
- Try refreshing the page or restarting server.
```

**Comment by souporserious** [comment]
Ok, I see what's going on here. The `mdxts/next` plugin starts the web socket server and doesn't detect it's running from the cli already. This shouldn't happen and they should work together so that's a bug. The directory not being created when using only the Next.js plugin is a bug as well.

I'll work on getting these both fixed up! Thank you for all of the details and testing 🙏

**Comment by souporserious** [comment]
The directory not being created is fixed now in the latest version `1.7.0`! As for the `EADDRINUSE` error, I've opted for only running either the CLI or Next.js plugin so the error message should be more detailed about this now. There currently isn't a reason to have both running so you should only need one or the other depending on if you want full customization or go the preconfigured route through the plugin.
